@hts-forking/src/slotmap.js

```diff
@@ -0,0 +1,158 @@
+/*-
+ * Hedera Hardhat Forking Plugin
+ *
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const { strict: assert } = require('assert');
+const { keccak256 } = require('ethers');
+
+const { toIntHex256, toSnakeCase } = require('./utils');
+const {
+    storageLayout: { storage, types },
+} = require('../resources/HtsSystemContract.json');
+
+class SlotMap {
+    constructor() {
+        /** @type {Map<bigint, {value: string, path: string, type: string}>} */
+        this._map = new Map();
+    }
+
+    /**
+     *
+     * @param {bigint} slot
+     * @param {string} value
+     * @param {string} path
+     * @param {string} type
+     */
+    store(slot, value, path, type) {
+        const prev = this._map.get(slot);
+        if (prev !== undefined)
+            throw new Error(`Slot \`${slot}\` in use by ${JSON.stringify(prev)}: ${value}@${path}`);
+        this._map.set(slot, { value: '0x' + value, path, type });
+    }
+}
+
+/**
+ * @type {{[t_name: string]: (value: string) => string[]}}
+ */
+const _types = {
+    t_string_storage: function (str) {
+        return this['t_bytes_storage'](Buffer.from(str).toString('hex'));
+    },
+    t_bytes_storage: hexstr => {
+        const len = Buffer.from(hexstr, 'hex').length;
+        const [header, lenByte, chunks] =
+            len > 31
+                ? [
+                      '0',
+                      len * 2 + 1,
+                      [...Array(Math.ceil(len / 32)).keys()].map(i =>
+                          hexstr.substring(i * 64, (i + 1) * 64).padEnd(64, '0')
+                      ),
+                  ]
+                : [hexstr, len * 2, []];
+
+        return [`${header.padEnd(64 - 2, '0')}${lenByte.toString(16).padStart(2, '0')}`, ...chunks];
+    },
+    t_uint8: val => [toIntHex256(val)],
+    t_uint256: val => [toIntHex256(val)],
+    t_int256: str => [toIntHex256(str ?? 0)],
+    t_address: str => [toIntHex256(str?.replace('0.0.', '') ?? 0)],
+    t_bool: value => [toIntHex256(value ? 1 : 0)],
+};
+
+/**
+ * @param {Record<string, unknown>} token
+ * @returns {Map<bigint, {value: string, path: string}>}
+ */
+function slotMapOf(token) {
```

Comment:
This method seems really hacky and tries to hit multiple rabbits with a single bullet. The `flatten` method also seems to have multiple purposes and is really hard to read and follow through.

Wouldn't it be better if we just constructed a new object that would represent the `TokenInfo` struct and mapped the mirror node response to it? 

For example, one way to do it can be by defining a new class that will represent the `TokenInfo` struct:
- in its constructor (or through a static method - e.g., `fromMirrorNodeToken`) it can be created from the mirror node object - something similar to what is done in `_initTokenInfo` in the foundry solution where the `TokenInfo` struct is being populated from the mirror node JSON response.

Then in order to fill the slots we would have to:
1. Construct an instance of the `TokenInfo` object, by populating it from the token response from mirror node
2. Iterate through the fields of the `TokenInfo` object and use the storage layout to figure out the slots where each field should be residing

This seems way easier to understand and also would be very similar to what is done in the `HtsSystemContractJson._initTokenInfo`  method

Reply:
Mmm, so _1._ and _2._ is essentially what `slotMapOf` is doing. The first bit corresponds loosely to _1._, but just amending the `token` so it's compatible with the Solidity definition. The second part is doing _2._, but driven by the storage layout instead.

`getHtsStorageAt` is driven by slot number (not fields), so to me it's more clear to follow the storage layout first.

Reply:
Okay, can we create a new object, instead of amending the token object? Also, in the new object we can directly write all fields explicitly in camelCase (as they are in the solidity struct), instead of snake case (this would skip the conversion logic). 

In this way it would be more explicit and clearer what fields exist in the mapped object and what we are doing, amending the token object is kind of ambiguous because we are only adding the missing fields and requires us to believe that the actual response will contain the rest of the fields, instead of catching any potential problems (missing fields, etc) on time.

Reply:
I wanted to avoid re-creating the whole object, just wanted to adapt what is needed in order to retrieve the field values. This way we reuse the storage layout (if we create a new object we are repeating ourselves by duplicating the storage layout).